### PR TITLE
Executable fork choice

### PR DESCRIPTION
**Pull request by CarlBeek**
This is a new PR which supersedes the old executable fork choice PR #1068. This was done because `dev` is so far ahead of the branch that #1068 was forked from so it was easier to start again.

Special thanks to @hwwhww and @vbuterin for the help tracking down a stupid bug.

### CODE REVIEW

**Comment by protolambda** [comment]
There's no good reason to make `Store` and other non-consensus data an SSZ object (`Container`). The typing is changing, and enforces proper ssz-types as elements. So this would break, and introducing python-dict support into ssz doesn't seem like the right solution. Instead just declare it as `class Store:` or `class Store(object)` (new python class style), and it'll work just fine.

**Comment by djrtwo** [comment]
hold up on my review before merge

File: specs/core/0_fork-choice.md [hunk 1]
```diff
@@ -34,76 +37,131 @@ All terminology, constants, functions, and protocol mechanics defined in the [Ph
...
 
-### Beacon chain fork choice rule
+#### `get_genesis_store`
 
-The beacon chain fork choice rule is a hybrid that combines justification and finality with Latest Message Driven (LMD) Greediest Heaviest Observed SubTree (GHOST). At any point in time, a validator `v` subjectively calculates the beacon chain head as follows.
+```python
+def get_genesis_store(genesis_state: BeaconState) -> Store:
+    genesis_block = BeaconBlock(state_root=hash_tree_root(genesis_state))
+    root = signing_root(genesis_block)
+    return Store(blocks={root: genesis_block}, states={root: genesis_state}, justified_root=root, finalized_root=root)
```
Comment (hwwhww) [inline specs/core/0_fork-choice.md]: Should we set `time=genesis_state.genesis_time`?
```suggestion
    return Store(
        blocks={root: genesis_block},
        states={root: genesis_state},
        time=genesis_state.genesis_time,
        justified_root=root, finalized_root=root,
    )
```

File: specs/core/0_fork-choice.md [hunk 2]
```diff
@@ -34,76 +37,131 @@ All terminology, constants, functions, and protocol mechanics defined in the [Ph
...
 
-During genesis, justification and finality root fields within the `BeaconState` reference `ZERO_HASH` rather than a known block. `ZERO_HASH` in `previous_justified_root`, `current_justified_root`, and `finalized_root` should be considered as an alias to the root of the genesis block.
+```python
+def on_block(store: Store, block: BeaconBlock) -> None:
+    # Add new block to the store
+    store.blocks[signing_root(block)] = block
+    # Check block is a descendant of the finalized block
+    assert get_ancestor(store, signing_root(block), store.blocks[store.finalized_root].slot) == store.finalized_root
+    # Check block slot against Unix time
+    pre_state = deepcopy(store.states[block.parent_root])
```
[the remark this answers is not in the document]
Comment (CarlBeek) [reply specs/core/0_fork-choice.md]: I think having `deepcopy` be explicit makes clearer what is actually happening, this proposed way hides what is happening in undocumented SSZ specifications. That said, I this is not a very strong preference of mine. (It's not like I've implemented a secret undocumented SSZ `Dict` type before or anything. 😇)

**Comment by djrtwo** [comment]
This needs a little bit of work wrt issues in #768.
I'm working on a PR to this PR :)

**Pull request merged**